From e6a8b9f0490b4ec428ee127e0eb9b81c30e46209 Mon Sep 17 00:00:00 2001
From: Patrick Combes <patrick.combes@intel.com>
Date: Thu, 19 Sep 2013 17:15:48 +0200
Subject: [PORT FROM MAIN][BT] Fix SIGV in bta_dm_queue_search function

BZ: 139604

Since GKI_getbuf can return NULL pointer in several cases, this pointer
should be tested before using it.

Change-Id: I09acf93b19648c6adef26e9c2adcb6d3dae8b0d8
Orig-Change-Id: I858f2a76839539345bf5c8544b8340999a768c94
Category: device enablement
Domain: CWS.BT-Common
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Patrick Combes <patrick.combes@intel.com>
Reviewed-on: http://android.intel.com:8080/133268
Reviewed-by: Gros, Julien <julien.gros@intel.com>
Tested-by: Lestienne, ChristopheX <christophex.lestienne@intel.com>
---
 bta/dm/bta_dm_act.c |   15 ++++++++-------
 1 files changed, 8 insertions(+), 7 deletions(-)
 mode change 100755 => 100644 bta/dm/bta_dm_act.c

diff --git a/bta/dm/bta_dm_act.c b/bta/dm/bta_dm_act.c
old mode 100755
new mode 100644
index 72f9544..0742624
--- a/bta/dm/bta_dm_act.c
+++ b/bta/dm/bta_dm_act.c
@@ -2042,8 +2042,8 @@ void bta_dm_queue_search (tBTA_DM_MSG *p_data)
         GKI_freebuf(bta_dm_search_cb.p_search_queue);
     }
 
-    bta_dm_search_cb.p_search_queue = (tBTA_DM_MSG *)GKI_getbuf(sizeof(tBTA_DM_API_SEARCH));
-    memcpy(bta_dm_search_cb.p_search_queue, p_data, sizeof(tBTA_DM_API_SEARCH));
+    if ((bta_dm_search_cb.p_search_queue = (tBTA_DM_MSG *)GKI_getbuf(sizeof(tBTA_DM_API_SEARCH))) != NULL)
+        memcpy(bta_dm_search_cb.p_search_queue, p_data, sizeof(tBTA_DM_API_SEARCH));
 
 }
 
@@ -2063,8 +2063,8 @@ void bta_dm_queue_disc (tBTA_DM_MSG *p_data)
         GKI_freebuf(bta_dm_search_cb.p_search_queue);
     }
 
-    bta_dm_search_cb.p_search_queue = (tBTA_DM_MSG *)GKI_getbuf(sizeof(tBTA_DM_API_DISCOVER));
-    memcpy(bta_dm_search_cb.p_search_queue, p_data, sizeof(tBTA_DM_API_DISCOVER));
+    if ((bta_dm_search_cb.p_search_queue = (tBTA_DM_MSG *)GKI_getbuf(sizeof(tBTA_DM_API_DISCOVER))) != NULL)
+        memcpy(bta_dm_search_cb.p_search_queue, p_data, sizeof(tBTA_DM_API_DISCOVER));
 
 }
 
@@ -5158,11 +5158,12 @@ static void bta_dm_gatt_disc_complete(UINT16 conn_id, tBTA_GATT_STATUS status)
             p_msg->disc_result.result.disc_res.device_type = BT_DEVICE_TYPE_BLE;
             if ( bta_dm_search_cb.ble_raw_used > 0 )
             {
-                p_msg->disc_result.result.disc_res.p_raw_data = GKI_getbuf(bta_dm_search_cb.ble_raw_used);
-
-                memcpy( p_msg->disc_result.result.disc_res.p_raw_data,
+                if ((p_msg->disc_result.result.disc_res.p_raw_data = GKI_getbuf(bta_dm_search_cb.ble_raw_used)) != NULL)
+                {
+                    memcpy( p_msg->disc_result.result.disc_res.p_raw_data,
                             bta_dm_search_cb.p_ble_rawdata,
                             bta_dm_search_cb.ble_raw_used );
+                }
 
                 p_msg->disc_result.result.disc_res.raw_data_size = bta_dm_search_cb.ble_raw_used;
             }
-- 
1.7.4.1

